### PR TITLE
feat(core) CDB-1630: Error on index query if sync hasn't completed

### PR DESCRIPTION
**Pull request by dbcfd**

### CODE REVIEW

**Comment by dbcfd** [comment]
@stephhuynh18 @stbrody This should be ready for final review now.

File: packages/core/src/indexing/database-index-api.ts [hunk 1]
```diff
@@ -201,17 +208,26 @@ export class DatabaseIndexApi<DateType = Date | number> {
...
+    if (foundModelToIndex == undefined) {
       const err = new Error(`Query failed: Model ${model} is not indexed on this node`)
       this.logger.debug(err)
       throw err
+    } else {
+      await this.assertModelIndexable(modelStreamId)
+    }
+  }
+
+  async assertModelIndexable(modelStreamId: StreamID | string): Promise<void> {
```
Comment (stbrody) [inline packages/core/src/indexing/database-index-api.ts:225]: I don't really understand the division in names between `assertModelQueryable` and `assertModelIndexable`.  This function only runs on models that are already being actively indexed and just checks whether queries should be allowed through or not, so seems weird to me to call it assertModelsIndexable when the caller (assertModelsQueryable) has already asserted that the model is actually actively indexed

File: packages/core/src/indexing/build-indexing.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import { type DatabaseIndexApi, PostgresIndexApi, SqliteIndexApi } from './datab
 import knex from 'knex'
 import { DiagnosticsLogger, Networks } from '@ceramicnetwork/common'
 import * as fs from 'fs'
+import { ISyncQueryApi } from '../sync/interfaces.js'
```
Comment (stephhuynh18) [inline packages/core/src/indexing/build-indexing.ts:5]: What is this used for?

File: packages/core/src/sync/interfaces.ts [hunk 3]
```diff
@@ -10,7 +10,11 @@ export type SyncConfig = {
   chainId: SupportedNetwork
 }
 
-export interface ISyncApi {
+export interface ISyncQueryApi {
```
Comment (stbrody) [inline packages/core/src/sync/interfaces.ts:13]: can we add some comments for what these interfaces represent?

**Review by stbrody — APPROVED** [review]
LGTM mod some nitpicks on naming.

Would still like @stephhuynh18 to check that the way `syncComplete` is checking whether the historical sync has completed is correct

File: packages/core/src/ceramic.ts [hunk 4]
```diff
@@ -302,6 +302,7 @@ export class Ceramic implements CeramicApi {
       this.repository.index,
       this._logger
     )
+    this.repository.index.setSyncQueryApi(this.syncApi)
```
Comment (stephhuynh18) [inline packages/core/src/ceramic.ts:305]: There is a way to get rid of the circular dependency: 

`SyncApi` only uses the indexApi to retrieve the indexable models. We can pass the indexable models in `SyncApi.init()`

File: packages/core/src/sync/sync-api.ts [hunk 5]
```diff
@@ -265,6 +265,10 @@ export class SyncApi implements ISyncApi {
     })
   }
 
+  async syncComplete(model: string): Promise<boolean> {
+    return !this.modelsToSync.has(model)
```
Comment (stephhuynh18) [inline packages/core/src/sync/sync-api.ts:269]: Unfortunately figuring out which models are currently syncing is a bit more complicated. There are two types of syncs: 1. continuous sync, 2. historic sync. 

We want to prevent queries if there are historic syncs going on for that model. The continuous sync is always running for all indexed models

File: packages/core/src/initialization/make-index-api.ts [hunk 6]
```diff
@@ -2,6 +2,7 @@ import type { DiagnosticsLogger } from '@ceramicnetwork/common'
 import { Networks } from '@ceramicnetwork/common'
 import { buildIndexing, IndexingConfig } from '../indexing/build-indexing.js'
 import { DatabaseIndexApi, SqliteIndexApi } from '../indexing/database-index-api.js'
+import { ISyncQueryApi } from '../sync/interfaces.js'
```
Comment (stephhuynh18) [inline packages/core/src/initialization/make-index-api.ts:5]: Where is this used?

File: packages/core/src/ceramic.ts (hunk 4 above)
Comment (stbrody) [reply packages/core/src/ceramic.ts:305]: ooh good call.  We'd also have to update it any time the set of indexed models changes, but that should be doable and would indeed resolve the circular dependency.  Probably fine to leave for a follow-up though?

File: packages/core/src/sync/sync-api.ts (hunk 5 above)
Comment (stbrody) [reply packages/core/src/sync/sync-api.ts:269]: Any ideas on how we could test this change effectively so a nuance like this would be caught?

Comment (stephhuynh18) [reply packages/core/src/sync/sync-api.ts:269]: We can check the active/pending jobs and see if the models is included, we can keep a running list of models that having pending history syncs and update when syncs are done. We can have a table with previous sync information for each model.

Comment (stephhuynh18) [reply packages/core/src/sync/sync-api.ts:269]: Easiest thing to do is to not allow ANY queries if history syncs are running. So basically if there are any active or pending history sync jobs

Comment (stbrody) [reply packages/core/src/sync/sync-api.ts:269]: I think disallowing all queries could cause issues.  Imagine an app has been running a while with some set of indexed models.  Now it wants to start using a new model, so it adds that model and starts the historical sync, but doesn't update the app logic to use the new model until the sync is complete.  But now the whole app will fail for the queries on all the existing models that were already working.

File: packages/core/src/indexing/build-indexing.ts (hunk 2 above)
Comment (dbcfd) [reply packages/core/src/indexing/build-indexing.ts:5]: Unused, should probably make that an error in lint. Originally was plumbing through the dependency.

File: packages/core/src/ceramic.ts (hunk 4 above)
Comment (dbcfd) [reply packages/core/src/ceramic.ts:305]: That actually only removes one of the circular dependencies. The main one is from 

`this.repository.stateManager.handleUpdate.bind(this.repository.stateManager),`

Since you have to setDeps on repository, which creates a stateManager, which needs a local index. And the local index would need a sync api.

File: packages/core/src/indexing/database-index-api.ts (hunk 1 above)
Comment (dbcfd) [reply packages/core/src/indexing/database-index-api.ts:223]: `assertModelQueryable` no longer calls `assertModelIndexable`. That along the lines of what you were thinking?

File: packages/core/src/sync/workers/sync.ts [hunk 7]
```diff
@@ -66,7 +67,7 @@ export class SyncWorker implements Worker<SyncJobData> {
    */
   async handler(job: PgBoss.Job) {
     const jobData = job.data as SyncJobData
-    const { fromBlock, toBlock, models } = jobData
+    const { fromBlock, toBlock, models, syncCompleteCallback } = jobData
```
Comment (stephhuynh18) [inline packages/core/src/sync/workers/sync.ts:70]: Interesting but unfortunately pg.boss serializes the data when its stored in postgres so syncCompleteCallback will not be a function but rather a stringified function. 

You could pass a callback to the worker itself

File: packages/core/src/indexing/database-index-api.ts (hunk 1 above)
Comment (stbrody) [reply packages/core/src/indexing/database-index-api.ts:223]: except it's still weird to call this `assertModelIndexable`.  It only runs on models that are already being indexed

File: packages/core/src/sync/sync-api.ts [hunk 8]
```diff
@@ -262,11 +266,16 @@ export class SyncApi implements ISyncApi {
       fromBlock: startBlock,
       toBlock: endBlock,
       models: modelIds,
+      syncCompleteCallback: this.modelSyncComplete,
     })
   }
 
+  modelSyncComplete(model: string) {
+    this.modelsToHistoricSync.delete(model)
```
Comment (stephhuynh18) [inline packages/core/src/sync/sync-api.ts:274]: a history sync can occur in the follow situations:
1. When a node is restart and needs to catch up
2. A node is syncing for the first time
3. A reorg happened and we need to re-sync X blocks
4. We add a new model

a sync is complete for a model when it does not have a sync running due to 2 or 4. I am not sure about 1 or 3.  A model can have multiple history syncs at once. Ex. A node restarts twice so it has two different catchup syncs. 

So basically I don't think modelsToHistoricSync should be a set, rather than a mapping with counts

File: packages/core/src/indexing/database-index-api.ts (hunk 1 above)
Comment (stbrody) [reply packages/core/src/indexing/database-index-api.ts:223]: maybe just call it `assertNoOngoingSyncs` and put it back into being called by `assertModelsQueryable`?

Comment (stbrody) [reply packages/core/src/indexing/database-index-api.ts:223]: you could even take the current body of `assertModelsQueryable` and move it to a new helper called `assertModelsAreIndexed`.  Then the body of `assertModelsQueryable` would just be 2 lines long, calling the two helper functions for the two checks that determine if a model is queryable (that it is indexed and that there are no ongoing syncs)

**Review by stbrody — COMMENTED** [review]
added comments on the helper function structure, but will defer to @stephhuynh18 on the correctness of the ongoing sync checks

File: packages/core/src/sync/sync-api.ts [hunk 9]
```diff
@@ -77,7 +82,7 @@ export interface SyncStatus {
 
 export class SyncApi implements ISyncApi {
   public readonly modelsToSync = new Set<string>()
-
+  public readonly modelsToHistoricSync = new Map<string, number>()
```
Comment (stephhuynh18) [inline packages/core/src/sync/sync-api.ts:85]: This needs to be incremented and decremented for every history sync job that prevents queries.  Right now its incremented only during `startModelSync` and decremented for ALL `history sync` (catch up, reorg, first time, and new model. This will result in the counts being messed up.

File: packages/core/src/sync/sync-api.ts (hunk 8 above)
Comment (stbrody) [reply packages/core/src/sync/sync-api.ts]: That's actually a really interesting question.  I think I agree with blocking cases 2 and 4 but letting queries go through in cases 1 and 3.  @oed @3avi @michaelsena for any other opinions.

File: packages/core/src/sync/sync-api.ts [hunk 10]
```diff
@@ -242,14 +265,17 @@ export class SyncApi implements ISyncApi {
...
     if (reorganized) {
       await this._addSyncJob(HISTORY_SYNC_JOB, {
+        jobType: SyncJobType.Reorg,
         fromBlock: block.number - BLOCK_CONFIRMATIONS,
         toBlock: block.number,
         models: Array.from(this.modelsToSync),
       })
     } else {
       await this._addSyncJob(CONTINUOUS_SYNC_JOB, {
+        jobType: SyncJobType.Reorg,
```
Comment (stephhuynh18) [inline packages/core/src/sync/sync-api.ts:278]: Not important but can we change this to "new block". This is the case when a new block is heard
